### PR TITLE
UCP/API: Flag to silence request leak check

### PR DESCRIPTION
**Pull request by yosefe**
# Why
Allow MPI runtime skip request leak check, which will silence errors like reported in https://github.com/open-mpi/ompi/issues/8426

### CODE REVIEW

**Comment by shamisp** [comment]
@yosefe - Why MPI_Finalize does not cleanup the resources ? AFAIK, It supposed to track communicators and clean the resources. Adding @hppritcha

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -165,11 +165,25 @@ enum ucp_worker_params_field {
...
                                                            descriptor */
+    UCP_WORKER_PARAM_FIELD_FLAGS        = UCS_BIT(5) /**< Worker flags */
 };
 
 
+/**
+ * @ingroup UCP_WORKER
+ * @brief UCP worker flags
+ *
+ * This enumeration allows pecifying flags for @ref ucp_worker_params_t.flags,
```
Comment (brminich) [inline src/ucp/api/ucp.h:178]: `specifying`

File: src/ucp/core/ucp_worker.c [hunk 2]
```diff
@@ -1981,7 +1982,8 @@ ucs_status_t ucp_worker_create(ucp_context_h context,
     }
 
     uct_thread_mode = UCS_THREAD_MODE_SINGLE;
-    worker->flags   = 0;
+    worker->flags   = (params->field_mask & UCP_WORKER_PARAM_FIELD_FLAGS) ?
```
Comment (brminich) [inline src/ucp/core/ucp_worker.c:1985]: minor: you can use `UCP_PARAM_VALUE`

File: src/ucp/core/ucp_worker.h [hunk 3]
```diff
@@ -59,9 +59,14 @@
  * UCP worker flags
  */
 enum {
-    UCP_WORKER_FLAG_EXTERNAL_EVENT_FD = UCS_BIT(0), /**< worker event fd is external */
-    UCP_WORKER_FLAG_EDGE_TRIGGERED    = UCS_BIT(1), /**< events are edge-triggered */
-    UCP_WORKER_FLAG_MT                = UCS_BIT(2)  /**< MT locking is required */
+    /** MT locking is required */
```
Comment (brminich) [inline src/ucp/core/ucp_worker.h:62]: are values moved to allow combining with new `params->flags`?
Can you please add a comment and some check they are not mixed?

**Comment by yosefe** [comment]
> @yosefe - Why MPI_Finalize does not cleanup the resources ? AFAIK, It supposed to track communicators and clean the resources. Adding @hppritcha

Some applications do not release all requests [even though required by MPI standard], and other MPI implementations do not warn about this. There are user applications that start to spew these warnings when ran with UCX, and users see this as UCX issue, and in cases just silence them by setting `UCX_LOG_LEVEL=error`

See similar reports: 
* https://github.com/open-mpi/ompi/issues/6912
* https://github.com/open-mpi/ompi/issues/6376
* https://github.com/openucx/ucx/issues/2777
* https://github.com/COSIMA/libaccessom2/issues/38

Since that kind of leak does would not usually have side effects (MPI_Finalize is executed right before process exit), it would be better to not print it in case of MPI.

**Comment by shamisp** [comment]
I think part of the annoyance here (from user perspective) is that it generates tons of errors... Would it be better to generate one-liner error: your application leaked NN requests.  @hppritcha @pavanbalaji is the preference from MPI community to silence this  ? thanks !

**Comment by pavanbalaji** [comment]
> I think part of the annoyance here (from user perspective) is that it generates tons of errors... Would it be better to generate one-liner error: your application leaked NN requests. @hppritcha @pavanbalaji is the preference from MPI community to silence this ? thanks !

MPICH only generates this when the user provides a special configure flag, not by default.  From MPI's perspective, it doesn't matter if you throw an error in UCX or not.  We keep track of MPI requests internally and would throw a warning (if the user specified the configure flag) anyway.  The UCX warning would be redundant in that case.

**Comment by shamisp** [comment]
> > I think part of the annoyance here (from user perspective) is that it generates tons of errors... Would it be better to generate one-liner error: your application leaked NN requests. @hppritcha @pavanbalaji is the preference from MPI community to silence this ? thanks !
> 
> MPICH only generates this when the user provides a special configure flag, not by default. From MPI's perspective, it doesn't matter if you throw an error in UCX or not. We keep track of MPI requests internally and would throw a warning (if the user specified the configure flag) anyway. The UCX warning would be redundant in that case.
> 
> We could also release the UCX requests internally in `MPI_Finalize` as you suggested. So it shouldn't matter one way or another for MPI libraries. But I'm not an MPI library developer anymore, so I can't speak for them. :-)

As I read - MPICH detects the leak but does not really clean it up.

**Review by brminich — APPROVED** [review]

**Review by hoopoepg — APPROVED** [review]

**Comment by hppritcha** [comment]
> I think part of the annoyance here (from user perspective) is that it generates tons of errors... Would it be better to generate one-liner error: your application leaked NN requests. @hppritcha @pavanbalaji is the preference from MPI community to silence this ? thanks !

preference for silence from the OMPI side.

**Comment by yosefe** [comment]
@shamisp are we good to go?

**Comment by shamisp** [comment]
👍  from my end. @tonycurtis can you please review documentation.

**Review by tonycurtis — CHANGES_REQUESTED** [review]
ucp.h:1137: could simplify to "if UCP_WORKER_......." not set, defaults to 0 ?

**Comment by yosefe** [comment]
test output:
```
[==========] Running 2 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 2 tests from all/test_ucp_worker_request_leak
[ RUN      ] all/test_ucp_worker_request_leak.tag_send_recv/0 <all/leak_check>
[     INFO ] send req: 0x1e0e040, recv req: 0x1e18080
[     INFO ] < object 0x1e0df40 was not returned to mpool ucp_requests >
[     INFO ] < object 0x1e17f80 was not returned to mpool ucp_requests >
[       OK ] all/test_ucp_worker_request_leak.tag_send_recv/0 (371 ms)
[ RUN      ] all/test_ucp_worker_request_leak.tag_send_recv/1 <all/leak_ignore>
[     INFO ] send req: 0x288b040, recv req: 0x2895080
[       OK ] all/test_ucp_worker_request_leak.tag_send_recv/1 (373 ms)
[----------] 2 tests from all/test_ucp_worker_request_leak (744 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test case ran. (744 ms total)
[  PASSED  ] 2 tests.
```

**Comment by yosefe** [comment]
@tonycurtis fixed the comment
@brminich can you pls review the gtest?

File: test/gtest/ucp/test_ucp_worker.cc [hunk 4]
```diff
@@ -462,3 +462,88 @@ UCS_TEST_P(test_ucp_worker_discard, wireup_ep_flush_ok_not_wait_comp) {
...
+        add_variant_with_value(variants, UCP_FEATURE_TAG, LEAK_IGNORE,
+                               "leak_ignore");
+    }
+
+    bool ignore_leak()
+    {
+        return get_variant_value(0) == LEAK_IGNORE;
+    }
+
+    /// @override
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_worker.cc:487]: minor extra `/` here and below

**Review by brminich — APPROVED** [review]

File: test/gtest/ucp/test_ucp_worker.cc (hunk 4 above)
Comment (yosefe) [reply test/gtest/ucp/test_ucp_worker.cc:487]: it's supposed to be doxgen comment
https://www.doxygen.nl/manual/docblocks.html

> A third alternative is to use a block of at least two C++ comment lines, where each line starts with an additional slash or an exclamation mark. Here are examples of the two cases:
> 
> ///
> /// ... text ...
> ///

Comment (brminich) [reply test/gtest/ucp/test_ucp_worker.cc:487]: aa, ok

**Review by tonycurtis — APPROVED** [review]

**Pull request merged**